pkg/gpu/stream.go

```diff
@@ -0,0 +1,178 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"math"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	gpuebpf "github.com/DataDog/datadog-agent/pkg/gpu/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// StreamHandler is responsible for receiving events from a single CUDA stream and generating
+// stats from them.
+type StreamHandler struct {
+	kernelLaunches []gpuebpf.CudaKernelLaunch
+	memAllocEvents map[uint64]gpuebpf.CudaMemEvent
+	kernelSpans    []*model.KernelSpan
+	allocations    []*model.MemoryAllocation
+	processEnded   bool // A marker to indicate that the process has ended, and this handler should be flushed
+}
+
+func newStreamHandler() *StreamHandler {
+	return &StreamHandler{
+		memAllocEvents: make(map[uint64]gpuebpf.CudaMemEvent),
+	}
+}
+
+func (sh *StreamHandler) handleKernelLaunch(event *gpuebpf.CudaKernelLaunch) {
+	// Copy events, as the memory can be overwritten in the ring buffer after the function returns
+	sh.kernelLaunches = append(sh.kernelLaunches, *event)
+}
+
+func (sh *StreamHandler) handleMemEvent(event *gpuebpf.CudaMemEvent) {
+	if event.Type == gpuebpf.CudaMemAlloc {
+		sh.memAllocEvents[event.Addr] = *event
+		return
+	}
+
+	alloc, ok := sh.memAllocEvents[event.Addr]
+	if !ok {
+		log.Warnf("Invalid free event: %v", event)
+		return
+	}
+
+	data := model.MemoryAllocation{
+		StartKtime: alloc.Header.Ktime_ns,
+		EndKtime:   event.Header.Ktime_ns,
+		Size:       alloc.Size,
+		IsLeaked:   false, // Came from a free event, so it's not a leak
+	}
+
+	sh.allocations = append(sh.allocations, &data)
+	delete(sh.memAllocEvents, event.Addr)
+}
+
+func (sh *StreamHandler) markSynchronization(ts uint64) {
+	span := sh.getCurrentKernelSpan(ts)
+	if span == nil {
+		return
+	}
+
+	sh.kernelSpans = append(sh.kernelSpans, span)
+
+	remainingLaunches := []gpuebpf.CudaKernelLaunch{}
+	for _, launch := range sh.kernelLaunches {
+		if launch.Header.Ktime_ns >= ts {
+			remainingLaunches = append(remainingLaunches, launch)
+		}
+	}
+	sh.kernelLaunches = remainingLaunches
+}
+
+func (sh *StreamHandler) handleSync(event *gpuebpf.CudaSync) {
+	// TODO: Worry about concurrent calls to this?
+	sh.markSynchronization(event.Header.Ktime_ns)
+}
+
+func (sh *StreamHandler) getCurrentKernelSpan(maxTime uint64) *model.KernelSpan {
+	span := model.KernelSpan{
+		StartKtime: math.MaxUint64,
+		EndKtime:   maxTime,
+		NumKernels: 0,
+	}
+
+	for _, launch := range sh.kernelLaunches {
+		if launch.Header.Ktime_ns >= maxTime {
```

Comment:
what does this mean?

Reply:
When a synchronization event happens, we want to get all launches before that event, not the ones after. Added a comment detailing it.